sqldelight-studio-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/SqlDelightVirtualFileListener.kt

```diff
@@ -15,11 +15,73 @@
  */
 package com.squareup.sqldelight.intellij
 
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
 import com.intellij.openapi.vfs.VirtualFileAdapter
 import com.intellij.openapi.vfs.VirtualFileEvent
+import com.intellij.openapi.vfs.VirtualFileMoveEvent
+import com.intellij.openapi.vfs.VirtualFilePropertyEvent
+import com.intellij.psi.PsiManager
+import com.squareup.sqldelight.SqliteCompiler
+import com.squareup.sqldelight.intellij.lang.SqlDelightFileViewProvider
+import com.squareup.sqldelight.intellij.lang.SqliteFile
+import java.util.LinkedHashMap
 
-class SqlDelightVirtualFileListener : VirtualFileAdapter() {
-  override fun beforeFileDeletion(event: VirtualFileEvent) {
+class SqlDelightVirtualFileListener(project: Project) : VirtualFileAdapter() {
+  // Before the event occurs we need to store a reference to the generated file. After the
+  // event has occurred the virtual file will have changed and will have a reference to the
+  // new generated file. However, we must perform any virtual file modifications in the
+  // fileDeleted/fileMoved functions because they have write access and will be bundled into a
+  // single action (which can be undo'd).
+  private val generatedFiles = LinkedHashMap<VirtualFile, VirtualFile?>()
+  private val VirtualFile.sqliteFile: SqliteFile?
+      get() = psiManager.findFile(this) as? SqliteFile
+  private val psiManager = PsiManager.getInstance(project)
+
+  private fun storeGeneratedFile(event: VirtualFileEvent) {
+    if (event.file.extension != SqliteCompiler.FILE_EXTENSION) return
+    generatedFiles.put(event.file, event.file.sqliteFile?.generatedVirtualFile)
+  }
+
+  private fun removeOldFile(event: VirtualFileEvent) {
+    if (event.file.extension != SqliteCompiler.FILE_EXTENSION) return
     SqlDelightManager.removeFile(event.file)
+    generatedFiles[event.file]?.delete(event.requestor)
+    generatedFiles.remove(event.file)
+  }
+
+  private fun generateNewFile(event: VirtualFileEvent) {
+    if (event.file.extension != SqliteCompiler.FILE_EXTENSION) return
+    (event.file.sqliteFile?.viewProvider as? SqlDelightFileViewProvider)?.generateJavaInterface(true)
```

Comment:
this is actually pretty cool. because `generateJavaInterface` happens synchronously, every write command action that it runs is part of this parent action. Because the virtual file is delegated, all the code that runs to create the initial generated virtual file is ALSO part of this parent action. What this means is if you rename a file from like `Test.sq` to `Test2.sq` it will delete the file `TestModel.java` and create `Test2Model.java`. And if you undo in intellij, it will both re-create `TestModel.java` and delete `Test2Model.java`. Very cool